File: backend/app/state/project_state.py
from typing import Dict, List, Optional
from app.models.types import TiltSeries, ScanConfig


class ProjectState:
    """
    Simple project state manager.
    Selections are managed by frontend - backend just stores current tilt series state.
    On save, frontend sends selections, backend writes directly to disk.
    """

    def __init__(self):
        self.config: Optional[ScanConfig] = None
        self.tilt_series: Dict[str, TiltSeries] = {}  # id -> TiltSeries

    def set_config(self, config: ScanConfig):
        self.config = config
        self.tilt_series.clear()

    def add_tilt_series(self, ts: TiltSeries):
        self.tilt_series[ts.id] = ts

    def get_tilt_series(self, ts_id: str) -> Optional[TiltSeries]:
        return self.tilt_series.get(ts_id)

    def remove_tilt_series_by_mdoc_path(self, mdoc_path: str):
        """Remove tilt series by mdoc path"""
        to_remove = [ts_id for ts_id, ts in self.tilt_series.items() if ts.mdocPath == mdoc_path]
        for ts_id in to_remove:
            del self.tilt_series[ts_id]

    def list_tilt_series(self) -> List[TiltSeries]:
        return list(self.tilt_series.values())

    def update_tilt_series_frames(self, mdoc_path: str, selections: Dict[int, bool]) -> bool:
        """
        Update tilt series frames based on selections (keep only selected frames).
        Returns True if found and updated.
        selections: zIndex -> selected (True = keep frame)
        """
        ts = next((t for t in self.tilt_series.values() if t.mdocPath == mdoc_path), None)
        if not ts:
            return False
        
        # Filter frames and update in one pass
        updated_frames = []
        min_angle = float('inf')
        max_angle = float('-inf')
        
        for frame in ts.frames:
            if selections.get(frame.zIndex, True):
                updated_frames.append(frame)
                if frame.angle < min_angle:
                    min_angle = frame.angle
                if frame.angle > max_angle:
                    max_angle = frame.angle
        
        if updated_frames:
            from app.models.types import TiltSeries
            updated_ts = TiltSeries(
                id=ts.id,
                mdocPath=ts.mdocPath,
                frames=updated_frames,
                angleRange=(min_angle, max_angle),
            )
            self.tilt_series[ts.id] = updated_ts
        
        return True


# Global state instance
project_state = ProjectState()

File: backend/src/state/project_state.rs
use std::collections::HashMap;
use std::sync::LazyLock;
use tokio::sync::RwLock;

use crate::models::types::{ScanConfig, TiltSeries};

/// Simple in-memory project state.
pub struct ProjectState {
    pub config: RwLock<Option<ScanConfig>>,
    pub tilt_series: RwLock<HashMap<String, TiltSeries>>,
}

impl ProjectState {
    pub fn new() -> Self {
        Self {
            config: RwLock::new(None),
            tilt_series: RwLock::new(HashMap::new()),
        }
    }

    pub async fn set_config(&self, config: ScanConfig) {
        let mut cfg = self.config.write().await;
        *cfg = Some(config);
        self.tilt_series.write().await.clear();
    }

    pub async fn add_tilt_series(&self, ts: TiltSeries) {
        self.tilt_series.write().await.insert(ts.id.clone(), ts);
    }

    pub async fn get_tilt_series(&self, ts_id: &str) -> Option<TiltSeries> {
        self.tilt_series.read().await.get(ts_id).cloned()
    }

    pub async fn list_tilt_series(&self) -> Vec<TiltSeries> {
        self.tilt_series.read().await.values().cloned().collect()
    }

    pub async fn remove_tilt_series_by_mdoc_path(&self, mdoc_path: &str) {
        let mut ts_map = self.tilt_series.write().await;
        ts_map.retain(|_, ts| ts.mdoc_path != mdoc_path);
    }

    pub async fn update_tilt_series_frames(
        &self,
        mdoc_path: &str,
        selections: &std::collections::HashMap<i32, bool>,
    ) -> Result<(), String> {
        let mut ts_map = self.tilt_series.write().await;
        let ts = ts_map.values().find(|ts| ts.mdoc_path == mdoc_path).cloned();

        if let Some(mut ts) = ts {
            let mut updated_frames = Vec::new();
            let mut min_angle = f64::MAX;
            let mut max_angle = f64::MIN;

            for frame in &ts.frames {
                if selections.get(&frame.z_index).copied().unwrap_or(true) {
                    if frame.angle < min_angle {
                        min_angle = frame.angle;
                    }
                    if frame.angle > max_angle {
                        max_angle = frame.angle;
                    }
                    updated_frames.push(frame.clone());
                }
            }

            if !updated_frames.is_empty() {
                ts.frames = updated_frames;
                ts.angle_range = if min_angle.is_finite() && max_angle.is_finite() {
                    (min_angle, max_angle)
                } else {
                    (0.0, 0.0)
                };
                ts_map.insert(ts.id.clone(), ts);
            } else {
                ts.frames = Vec::new();
                ts_map.insert(ts.id.clone(), ts);
            }
            Ok(())
        } else {
            Err(format!("tilt series not found for mdoc path: {mdoc_path}"))
        }
    }
}

/// Global project state instance
pub static PROJECT_STATE: LazyLock<ProjectState> = LazyLock::new(ProjectState::new);
